src/mlpack/core/optimizers/fw/frank_wolfe.hpp

```diff
@@ -0,0 +1,177 @@
+ *
+ */
+template<
+    typename FunctionType,
+    typename LinearConstrSolverType,
+    typename UpdateRuleType
+>
+class FrankWolfe
+{
+ public:
+  /**
+   * Construct the Frank-Wolfe optimizer with the given function and
+   * parameters. Notice that the constraint domain \f$ D \f$ is input
+   * at the initialization of linear_constr_solver, the function to be  
+   * optimized is stored in update_rule.
+   *
+   * @param function Function to be optimized
+   * @param linear_constr_solver Solver for linear constrained problem.
+   * @param update_rule Rule for updating solution in each iteration.
+   * @param maxIterations Maximum number of iterations allowed (0 means no
+   *     limit).
+   * @param tolerance Maximum absolute tolerance to terminate algorithm.
+   */
+  FrankWolfe(FunctionType& function,
+    const LinearConstrSolverType linear_constr_solver,
+        const UpdateRuleType update_rule,
+        const size_t maxIterations = 100000,
+        const double tolerance = 1e-10);
+
+  /**
+   * Optimize the given function using FrankWolfe.  The given starting
+   * point will be modified to store the finishing point of the algorithm, and
+   * the final objective value is returned.
+   *
+   * @param iterate Starting point (will be modified).
+   * @return Objective value of the final point.
+   */
+  double Optimize(arma::mat& iterate);
+
+  //! Get the instantiated function to be optimized.
+  const FunctionType& Function() const { return function; }
+  //! Modify the instantiated function.
+  FunctionType& Function() { return function; }
+
+  //! Get the linear constrained solver.
+  LinearConstrSolverType LinearConstrSolver()
+      const { return linear_constr_solver; }
+  //! Modify the linear constrained solver.
+  LinearConstrSolverType& LinearConstrSolver() { return linear_constr_solver; }
+
+  //! Get the update rule.
+  UpdateRuleType UpdateRule() const { return update_rule; }
+  //! Modify the update rule.
+  UpdateRuleType& UpdateRule() { return update_rule; }
+
+  //! Get the maximum number of iterations (0 indicates no limit).
+  size_t MaxIterations() const { return maxIterations; }
+  //! Modify the maximum number of iterations (0 indicates no limit).
+  size_t& MaxIterations() { return maxIterations; }
+
+  //! Get the tolerance for termination.
+  double Tolerance() const { return tolerance; }
+  //! Modify the tolerance for termination.
+  double& Tolerance() { return tolerance; }
+
+ private:
+  //! The instantiated function.
+  FunctionType& function;
+
+  //! The solver for constrained linear problem in first step.
+  LinearConstrSolverType linear_constr_solver;
+
+  //! The rule to update, used in the second step.
+  UpdateRuleType update_rule;
+
+  //! The maximum number of allowed iterations.
+  size_t maxIterations;
+
+  //! The tolerance for termination.
+  double tolerance;
+};
+
+//! Orthogonal Matching Pursuit
+using OMP = FrankWolfe<FuncSq, ConstrLpBallSolver, UpdateSpan<FuncSq>>;
```

Comment:
If this is only used in the tests, I think we should avoid giving a typedef here, unless we plan to use it somewhere else in the code.